File: src/lua/graphics/Window.lua
-- https://github.com/Archzenvald/Shido
-- MIT license (see LICENSE, src/c/core.h or src/lua/core.lua)

local ffi = require("ffi")
local API = require("shido.API")
local core = require("shido.core")

-- C API
API.defineHandle("Window")
API.defineHandle("SwapChain")
ffi.cdef[[
shido_WindowRef* shido_Window_new(const char *title, int w, int h);
shido_WindowRef* shido_Window_get(uint32_t id);
int shido_Window_getDisplayIndex(shido_Window *self);
bool shido_Window_hasInputGrab(shido_Window *self);
void shido_Window_getMinimumSize(shido_Window *self, int out[2]);
void shido_Window_getMaximumSize(shido_Window *self, int out[2]);
float shido_Window_getOpacity(shido_Window *self);
void shido_Window_getPosition(shido_Window *self, int out[2]);
void shido_Window_getSize(shido_Window *self, int out[2]);
const char* shido_Window_getTitle(shido_Window *self);
void shido_Window_hide(shido_Window *self);
void shido_Window_show(shido_Window *self);
void shido_Window_maximize(shido_Window *self);
void shido_Window_minimize(shido_Window *self);
void shido_Window_raise(shido_Window *self);
void shido_Window_restore(shido_Window *self);
bool shido_Window_setFullscreen(shido_Window *self, bool enable, bool desktop);
bool shido_Window_setDisplayMode(shido_Window *self, const SDL_DisplayMode *mode);
void shido_Window_setInputGrab(shido_Window *self, bool enable);
void shido_Window_setResizable(shido_Window *self, bool flag);
void shido_Window_setBordered(shido_Window *self, bool flag);
void shido_Window_setOpacity(shido_Window *self, float opacity);
void shido_Window_setMinimumSize(shido_Window *self, int w, int h);
void shido_Window_setMaximumSize(shido_Window *self, int w, int h);
void shido_Window_setSize(shido_Window *self, int w, int h);
void shido_Window_setPosition(shido_Window *self, int x, int y);
void shido_Window_setTitle(shido_Window *self, const char *title);
void shido_Window_getFullscreen(shido_Window *self, bool out[2]);
bool shido_Window_isVisible(shido_Window *self);
bool shido_Window_isResizable(shido_Window *self);
bool shido_Window_isBordered(shido_Window *self);
bool shido_Window_isMaximized(shido_Window *self);
bool shido_Window_isMinimized(shido_Window *self);
bool shido_Window_hasFocus(shido_Window *self);
bool shido_Window_hasMouseFocus(shido_Window *self);

shido_SwapChainRef* shido_SwapChain_new(shido_WindowRef *window);
]]
local L = ffi.load("shido")

local graphics = shido.graphics

-- Window

function graphics.newWindow(title, w, h)
  local ref = L.shido_Window_new(title, w, h)
  if ref == nil then core.error() end
  return API.claimFreeRef(ref)
end

-- Get window from id.
function graphics.getWindow(id)
  local ref = L.shido_Window_get(id)
  if ref ~= nil then return API.claimFreeRef(ref) end
end

-- methods
local Window = {}

-- Get window display index.
-- return 1-based index
function Window:getDisplayIndex()
  local index = L.shido_Window_getDisplayIndex(self.obj)
  if index < 0 then core.error() end
  return index+1
end
function Window:hasInputGrab() return L.shido_Window_hasInputGrab(self.obj) end

function Window:getMinimumSize()
  local out = ffi.new("int[2]")
  L.shido_Window_getMinimumSize(self.obj, out)
  return out[0], out[1]
end
function Window:getMaximumSize()
  local out = ffi.new("int[2]")
  L.shido_Window_getMaximumSize(self.obj, out)
  return out[0], out[1]
end
function Window:getOpacity() return L.shido_Window_getOpacity(self.obj) end
function Window:getPosition()
  local out = ffi.new("int[2]")
  L.shido_Window_getPosition(self.obj, out)
  return out[0], out[1]
end
function Window:getSize()
  local out = ffi.new("int[2]")
  L.shido_Window_getSize(self.obj, out)
  return out[0], out[1]
end
function Window:getTitle() return ffi.string(L.shido_Window_getTitle(self.obj)) end
function Window:hide() L.shido_Window_hide(self.obj) end
function Window:show() L.shido_Window_show(self.obj) end
function Window:maximize() L.shido_Window_maximize(self.obj) end
function Window:minimize() L.shido_Window_minimize(self.obj) end
function Window:raise() L.shido_Window_raise(self.obj) end
function Window:restore() L.shido_Window_restore(self.obj) end
-- Set fullscreen state.
-- desktop: (optional) true for non-exclusive desktop/windowed fullscreen
-- return true or (false, err) on failure
function Window:setFullscreen(enable, desktop)
  local ok = L.shido_Window_setFullscreen(self.obj, enable, not not desktop)
  if not ok then return ok, core.getError() end
  return ok
end
-- Set fullscreen display mode.
-- mode: as returned by graphics.getDisplayMode()
-- return true or (false, err) on failure
function Window:setDisplayMode(mode)
  local ok = L.shido_Window_setDisplayMode(self.obj, mode.raw)
  if not ok then return ok, core.getError() end
  return ok
end

function Window:setInputGrab(enable) L.shido_Window_setInputGrab(self.obj, enable) end
function Window:setResizable(flag) L.shido_Window_setResizable(self.obj, flag) end
function Window:setBordered(flag) L.shido_Window_setBordered(self.obj, flag) end
function Window:setOpacity(opacity) L.shido_Window_setOpacity(self.obj, opacity) end
function Window:setMinimumSize(w, h) L.shido_Window_setMinimumSize(self.obj, w, h) end
function Window:setMaximumSize(w, h) L.shido_Window_setMaximumSize(self.obj, w, h) end
function Window:setSize(w, h) L.shido_Window_setSize(self.obj, w, h) end
function Window:setPosition(x, y) L.shido_Window_setPosition(self.obj, x, y) end
function Window:setTitle(title) L.shido_Window_setTitle(self.obj, title) end

function Window:getFullscreen()
  local out = ffi.new("bool[2]")
  L.shido_Window_getFullscreen(self.obj, out)
  return out[0], out[1]
end
function Window:isVisible() return L.shido_Window_isVisible(self.obj) end
function Window:isResizable() return L.shido_Window_isResizable(self.obj) end
function Window:isBordered() return L.shido_Window_isBordered(self.obj) end
function Window:isMaximized() return L.shido_Window_isMaximized(self.obj) end
function Window:isMinimized() return L.shido_Window_isMinimized(self.obj) end
function Window:hasFocus() return L.shido_Window_hasFocus(self.obj) end
function Window:hasMouseFocus() return L.shido_Window_hasMouseFocus(self.obj) end

API.implementHandle("Window", Window)

-- SwapChain

function graphics.newSwapChain(window)
  local ref = L.shido_SwapChain_new(window.ref)
  if ref == nil then core.error() end
  return API.claimFreeRef(ref)
end

-- methods
local SwapChain = {}

API.implementHandle("SwapChain", SwapChain)

File: src/c/graphics/Window.cpp
// https://github.com/Archzenvald/Shido
// MIT license (see LICENSE, src/c/core.h or src/lua/core.lua)

#include "Window.hpp"

SHIDO_HANDLE_CPP(Window)
SHIDO_HANDLE_CPP(SwapChain)

// Window

shido::Registry<SDL_Window*, shido_WindowRef>& shido_Window::getRegistry()
{
  static shido::Registry<SDL_Window*, shido_WindowRef> registry;
  return registry;
}

shido_Window::~shido_Window()
{
  shido_Window::getRegistry().unreg(handle);
  SDL_DestroyWindow(handle);
}

shido_WindowRef* shido_Window_new(const char *title, int w, int h)
{
  SDL_Window *handle = SDL_CreateWindow(title, SDL_WINDOWPOS_CENTERED,
    SDL_WINDOWPOS_CENTERED, w, h, SDL_WINDOW_RESIZABLE);
  if(!handle){
    shido::error(std::string("SDL: ")+SDL_GetError());
    return nullptr;
  }
  shido_WindowRef *ref = new shido_WindowRef{std::make_shared<shido_Window>(handle)};
  shido_Window::getRegistry().reg(handle, ref);
  return ref;
}

shido_WindowRef* shido_Window_get(uint32_t id)
{ return shido_Window::getRegistry().get(SDL_GetWindowFromID(id)); }

int shido_Window_getDisplayIndex(shido_Window *self)
{
  int index = SDL_GetWindowDisplayIndex(self->handle);
  if(index < 0) shido::error(std::string("SDL: ")+SDL_GetError());
  return index;
}

bool shido_Window_hasInputGrab(shido_Window *self)
{ return SDL_GetWindowGrab(self->handle); }

void shido_Window_getMinimumSize(shido_Window *self, int out[2])
{ SDL_GetWindowMinimumSize(self->handle, out, out+1); }

void shido_Window_getMaximumSize(shido_Window *self, int out[2])
{ SDL_GetWindowMaximumSize(self->handle, out, out+1); }

float shido_Window_getOpacity(shido_Window *self)
{
  float out = 1.0f;
  SDL_GetWindowOpacity(self->handle, &out);
  return out;
}

void shido_Window_getPosition(shido_Window *self, int out[2])
{ SDL_GetWindowPosition(self->handle, out, out+1); }

void shido_Window_getSize(shido_Window *self, int out[2])
{ SDL_GetWindowSize(self->handle, out, out+1); }

const char* shido_Window_getTitle(shido_Window *self)
{ return SDL_GetWindowTitle(self->handle); }

void shido_Window_hide(shido_Window *self){ SDL_HideWindow(self->handle); }
void shido_Window_show(shido_Window *self){ SDL_ShowWindow(self->handle); }
void shido_Window_maximize(shido_Window *self){ SDL_MaximizeWindow(self->handle); }
void shido_Window_minimize(shido_Window *self){ SDL_MinimizeWindow(self->handle); }
void shido_Window_raise(shido_Window *self){ SDL_RaiseWindow(self->handle); }
void shido_Window_restore(shido_Window *self){ SDL_RestoreWindow(self->handle); }

bool shido_Window_setFullscreen(shido_Window *self, bool enable, bool desktop)
{
  Uint32 flags = 0;
  if(enable)
    flags = (desktop ? SDL_WINDOW_FULLSCREEN_DESKTOP : SDL_WINDOW_FULLSCREEN);
  bool ok = (SDL_SetWindowFullscreen(self->handle, flags) == 0);
  if(!ok)
    shido::error(std::string("SDL: ")+SDL_GetError());
  return ok;
}

bool shido_Window_setDisplayMode(shido_Window *self, const SDL_DisplayMode *mode)
{
  bool ok = (SDL_SetWindowDisplayMode(self->handle, mode) == 0);
  if(!ok)
    shido::error(std::string("SDL: ")+SDL_GetError());
  return ok;
}

void shido_Window_setInputGrab(shido_Window *self, bool enable)
{ SDL_SetWindowGrab(self->handle, enable ? SDL_TRUE : SDL_FALSE); }
void shido_Window_setResizable(shido_Window *self, bool flag)
{ SDL_SetWindowResizable(self->handle, flag ? SDL_TRUE : SDL_FALSE); }
void shido_Window_setBordered(shido_Window *self, bool flag)
{ SDL_SetWindowBordered(self->handle, flag ? SDL_TRUE : SDL_FALSE); }
void shido_Window_setOpacity(shido_Window *self, float opacity)
{ SDL_SetWindowOpacity(self->handle, opacity); }
void shido_Window_setMinimumSize(shido_Window *self, int w, int h)
{ SDL_SetWindowMinimumSize(self->handle, w, h); }
void shido_Window_setMaximumSize(shido_Window *self, int w, int h)
{ SDL_SetWindowMaximumSize(self->handle, w, h); }
void shido_Window_setSize(shido_Window *self, int w, int h)
{ SDL_SetWindowSize(self->handle, w, h); }
void shido_Window_setPosition(shido_Window *self, int x, int y)
{ SDL_SetWindowPosition(self->handle, x, y); }
void shido_Window_setTitle(shido_Window *self, const char *title)
{ SDL_SetWindowTitle(self->handle, title); }

void shido_Window_getFullscreen(shido_Window *self, bool out[2])
{
  Uint32 flags = SDL_GetWindowFlags(self->handle);
  out[0] = (flags & SDL_WINDOW_FULLSCREEN) || (flags & SDL_WINDOW_FULLSCREEN_DESKTOP);
  out[1] = flags & SDL_WINDOW_FULLSCREEN_DESKTOP;
}
bool shido_Window_isVisible(shido_Window *self)
{ return SDL_GetWindowFlags(self->handle) & SDL_WINDOW_SHOWN; }
bool shido_Window_isResizable(shido_Window *self)
{ return SDL_GetWindowFlags(self->handle) & SDL_WINDOW_RESIZABLE; }
bool shido_Window_isBordered(shido_Window *self)
{ return !(SDL_GetWindowFlags(self->handle) & SDL_WINDOW_BORDERLESS); }
bool shido_Window_isMaximized(shido_Window *self)
{ return SDL_GetWindowFlags(self->handle) & SDL_WINDOW_MAXIMIZED; }
bool shido_Window_isMinimized(shido_Window *self)
{ return SDL_GetWindowFlags(self->handle) & SDL_WINDOW_MINIMIZED; }
bool shido_Window_hasFocus(shido_Window *self)
{ return SDL_GetWindowFlags(self->handle) & SDL_WINDOW_INPUT_FOCUS; }
bool shido_Window_hasMouseFocus(shido_Window *self)
{ return SDL_GetWindowFlags(self->handle) & SDL_WINDOW_MOUSE_FOCUS; }

// SwapChain

shido_SwapChain::~shido_SwapChain()
{
  shido::Graphics::get().engine->destroy(handle);
}

shido_SwapChainRef* shido_SwapChain_new(shido_WindowRef *window)
{
  // get native window handle
  /// info
  void *whandle = nullptr;
  SDL_SysWMinfo winfo;
  SDL_VERSION(&winfo.version);
  if(!SDL_GetWindowWMInfo(window->ref->handle, &winfo)){
    shido::error(std::string("SDL: ")+SDL_GetError());
    return nullptr;
  }
  /// handle
  switch(winfo.subsystem){
#ifdef SDL_VIDEO_DRIVER_WINDOWS
    case SDL_SYSWM_WINDOWS: whandle = (void*)winfo.info.win.window; break;
#endif
#ifdef SDL_VIDEO_DRIVER_X11
    case SDL_SYSWM_X11: whandle = (void*)winfo.info.x11.window; break;
#endif
    default: break;
  }
  if(!whandle){
    shido::error("Native window handle not supported.");
    return nullptr;
  }
  shido::Graphics &graphics = shido::Graphics::get();
  filament::SwapChain *handle = graphics.engine->createSwapChain(whandle);
  if(!handle){
    shido::error("Filament SwapChain creation failed.");
    return nullptr;
  }
  return new shido_SwapChainRef{std::make_shared<shido_SwapChain>(window, handle)};
}
